connectors/schemaadapter/schema_reducer.go

```diff
@@ -0,0 +1,171 @@
+package schemaadapter
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/uber-go/dosa"
+	"github.com/uber-go/dosa/connectors/base"
+)
+
+// NewConnector initializes a Schema Reducer Connector
+func NewConnector(baseConnector dosa.Connector) dosa.Connector {
+	return &Connector{baseConnector}
+}
+
+// Connector for schema transformer
+type Connector struct {
+	// TODO why base connector? Why not just a dosa connector?
+	base.Connector
+}
+
+// MultiRead TODO
+func (c *Connector) MultiRead(ctx context.Context, ei *dosa.EntityInfo, values []map[string]dosa.FieldValue, minimumFields []string) ([]*dosa.FieldValuesOrError, error) {
+	return nil, nil
+}
+
+// Upsert calls Next
+func (c *Connector) Upsert(ctx context.Context, ei *dosa.EntityInfo, values map[string]dosa.FieldValue) error {
+	// Get all the partition keys and their values
+	keysMap := map[string]dosa.FieldValue{}
+	for pk := range ei.Def.PartitionKeySet() {
+		if value, ok := values[pk]; ok {
+			keysMap[pk] = value
+		}
+	}
+
+	key, _ := json.Marshal(keysMap)
+	value, _ := json.Marshal(values)
+	eiCopy := adaptToKeyValue(ei)
+	newValues := map[string]dosa.FieldValue {
+		"key": key,
+		"value": value,
+	}
+	return c.Connector.Upsert(ctx, eiCopy, newValues)
+}
+
+// MultiUpsert TODO
+func (c *Connector) MultiUpsert(ctx context.Context, ei *dosa.EntityInfo, values []map[string]dosa.FieldValue) ([]error, error) {
+	return nil, nil
+}
+
+// Remove removes a key
+func (c *Connector) Remove(ctx context.Context, ei *dosa.EntityInfo, values map[string]dosa.FieldValue) error {
+	key, _ := json.Marshal(values)
+	eiCopy := adaptToKeyValue(ei)
+
+	newValues := map[string]dosa.FieldValue{
+		"key": key,
+	}
+	return c.Connector.Remove(ctx, eiCopy, newValues)
+}
+
+// Range calls Next
+func (c *Connector) Range(ctx context.Context, ei *dosa.EntityInfo, columnConditions map[string][]*dosa.Condition, minimumFields []string, token string, limit int) ([]map[string]dosa.FieldValue, string, error) {
+	// TODO serializing dosa.Condition array? conditions could be any order
+
+	keysMap := map[string]interface{} {
+		"conditions": columnConditions,
+		"token": token,
+		"limit": limit,
+	}
+	key, _ := json.Marshal(keysMap)
+
+	eiCopy := adaptToKeyValue(ei)
+	newValues := map[string]dosa.FieldValue{
+		"key": key,
+	}
+
+	response, err :=  c.Connector.Read(ctx, eiCopy, newValues, minimumFields)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// TODO move this to global
+	type rangeResults struct {
+		token string
+		rows []map[string]dosa.FieldValue
+	}
+
+	unpack := rangeResults{}
+	err = json.Unmarshal(response, &unpack)
+	return unpack.rows, unpack.token, err
+
+	// TODO who is supposed to be calling the redis connector Upsert with the range response adapted?
```

Comment:
It seems like the fallback cache connector has to take care of most of this adaptation logic, not this schema reducing connector. 

For example, for this range method, the fallback cache connector takes the range response, creates a key for it, and either calls the redis connector's upsert or the redis connector's read. I no longer see how this SchemaReducing connector fits into the flow.